api/app/signals/apps/questionnaires/services/session.py

```diff
@@ -0,0 +1,234 @@
+                unanswered_by_id[question.id] = question
+            if answer is not None:
+                reachable_answers_by_question_id[answer.question.id] = answer
+
+            previous_question = question
+            answer_payload = None if answer is None else answer.payload
+            question = SessionService._get_next_question(nx_graph, questions_by_id, question, answer_payload)
+
+        # Finally we want to know whether session under consideration can be
+        # frozen meaningfully (i.e. a path through it is fully answered and the
+        # data can be made available for further processing).
+        can_freeze = False
+        if nx_graph.out_degree(previous_question.id) == 0 and not unanswered_by_id:
+            # Endpoint reached (no outgoing edges) and no unanswered required questions.
+            can_freeze = True
+
+        return reachable_questions_by_id, unanswered_by_id, reachable_answers_by_question_id, can_freeze
+
+    @staticmethod
+    def _get_endpoint_questions(nx_graph, questions_by_id):
+        """
+        Get endpoint questions in QuestionGraph.
+        """
+        endpoint_questions_by_id = {}
+        for question_id, out_degree in nx_graph.out_degree():
+            if out_degree == 0:
+                endpoint_questions_by_id[question_id] = questions_by_id[question_id]
+
+        return endpoint_questions_by_id
+
+    def get_next_question(self, question, answer):
+        """
+        Given question and its answer determine next question.
+        """
+        return SessionService._get_next_question(
+            self.question_graph_service.nx_graph,
+            self.question_graph_service.questions_by_id,
+            question,
+            answer.payload
+        )
+
+    def get_answers_by_analysis_key(self):
+        """
+        Given (partially?) answered graph determine answers along current path.
+        """
+        # translate answers along path to dict keyed by question.analysis_key
+        return {a.question.analysis_key: a for a in self.path_answers_by_question_id.values()}
+
+    def get_answers_by_uuid(self):
+        return {a.question.uuid: a for a in self.path_answers_by_question_id.values()}
+
+    def get_extra_properties(self, category_url=None):
+        """
+        Given (partially?) answered graph determine answers along current path
+        and return them in an Signal.extra_properties compatible dictionary.
+        """
+        # TODO: double check against VNG forms implementation
+        # Translate answers along path to dict keyed by question.analysis_key
+        # TBD, still how do we mark answers with the proper category
+        extra_props = []
+        for answer in self.path_answers_by_question_id.values():
+            entry = {
+                'id': answer.question.analysis_key,
+                'label': answer.question.short_label,
+                'answer': answer.payload,
+            }
+            if category_url:
```

Comment:
This `get_extra_properties` is a first cut at the problem to get the conversation started. What I had in mind was that we will want to support both old-style `extra_properties` blobs (especially during the transition to questionnaires app) and `extra_properties` extracted from questionnaires.

To construct `extra_properties` objects in our API I would:
1. If present start with a copy of `signal.extra_properties`.
2. Then move through any frozen `Session`s associated with our `Signal` (old to new) and add any extra property we find to the `extra_properties` (overwriting if we must).

This needs embellishment I am sure, especially in step 2. We may have to treat Sessions differently based on their `questionnaire.flow` property and need rules for how we determine the associated category. Such a rule could for instance be: use the category that was set at the time the questionnaire was answered.

I think this would allow a more general reaction request flow in the future where possibly whole questionnaires are asked, not just single questions.

Having said all that, as you mention, this is just a start.